examples/c++-api/AtomSpaceEventSubscribeExample.cc

```diff
@@ -0,0 +1,125 @@
+#include <iostream>
+
+#include <boost/signals2.hpp>
+
+#include <opencog/atomspace/AtomSpace.h>
+#include <opencog/truthvalue/AttentionValue.h>
+#include <opencog/truthvalue/SimpleTruthValue.h>
+#include <opencog/truthvalue/TruthValue.h>
+
+using namespace opencog;
+
+/**
+ * Sample code that shows how to listen for AtomSpace related events.
+ * The followings are all the events/signals supported by the AtomSpace.
+ */
+
+void AtomAddedCBHandler(const Handle& h);
+void AVChangedCBHandler(const Handle& h, const AttentionValuePtr& av_old,
+                        const AttentionValuePtr& av_new);
+void TVChangedCBHandler(const Handle& h, const TruthValuePtr& av_old,
+                        const TruthValuePtr& tv_new);
+void AtomRemovedCBHandler(const AtomPtr&);
+void AtomAddedToAFCBHandler(const Handle& h, const AttentionValuePtr& av_old,
+                            const AttentionValuePtr& av_new);
+void AtomRemovedFromAFCBHandler(const Handle& h,
+                                const AttentionValuePtr& av_old,
+                                const AttentionValuePtr& av_new);
+
+int main(int argc, char ** args)
+{
+    AtomSpace as;
+
+    boost::signals2::connection AtomAddedSignalConn,
+                                AVChangedSignalConn,
+                                TVChangedSignalConn,
+                                AtomsRemovedSignalConn,
+                                AtomAddedToAttentionalFocusSignalConn,
+                                AtomRemovedFromAttentionalFocusSignalConn;
+
+    //Register Atomspace Event call back handlers.
+    AtomAddedSignalConn = as.addAtomSignal(
+            boost::bind(&AtomAddedCBHandler, _1));
+    AVChangedSignalConn = as.AVChangedSignal(
+            boost::bind(&AVChangedCBHandler, _1, _2, _3));
+    TVChangedSignalConn = as.TVChangedSignal(
+            boost::bind(&TVChangedCBHandler, _1, _2, _3));
+    AtomsRemovedSignalConn = as.removeAtomSignal(
+            boost::bind(&AtomRemovedCBHandler, _1));
+    AtomAddedToAttentionalFocusSignalConn = as.AddAFSignal(
+            boost::bind(&AtomAddedToAFCBHandler, _1, _2, _3));
+    AtomRemovedFromAttentionalFocusSignalConn = as.RemoveAFSignal(
+            boost::bind(&AtomRemovedFromAFCBHandler, _1, _2, _3));
```

Comment:
I agree on most of the issues you have raised. However, these signals seem to be useful in some cases; For instance, the Hebbian creation part of ECAN mainly relies on atom added signal. I wonder if there is some other better way to implement signals in the atomspace rather than discouraging their use. Probably by generalizing the approach take in the [HebbianCreation code](https://github.com/opencog/opencog/blob/master/opencog/attention/HebbianCreationModule.cc#L77) which uses separate thread for processing the signal.